From 3d1f6102cde91d1dea6220955d600929c8bc40ba Mon Sep 17 00:00:00 2001
From: Richard Zhao <b20223@freescale.com>
Date: Tue, 17 Feb 2009 19:43:38 +0800
Subject: [PATCH] ENGR00107806-4 imx51: sgtl5000: use func to control spk amp shutdown

Becuase spk amp shutdown pin can be connected to PM gpo or cpu GPIO, I use a
function amp_enable in arch/arm/mach-xxx/mxxxx-3stack.c to controll shutdown.
This is imx51 specific modification.

Signed-off-by: Richard Zhao <b20223@freescale.com>
---
 arch/arm/mach-mx51/mx51_3stack.c |   42 +++++++++++++++++++++++++++++++++++++-
 1 files changed, 41 insertions(+), 1 deletions(-)

diff --git a/arch/arm/mach-mx51/mx51_3stack.c b/arch/arm/mach-mx51/mx51_3stack.c
index f00baf0..d6266e0 100644
--- a/arch/arm/mach-mx51/mx51_3stack.c
+++ b/arch/arm/mach-mx51/mx51_3stack.c
@@ -789,6 +789,9 @@ static void __init mxc_init_audio(void)
 
 #if defined(CONFIG_SND_SOC_IMX_3STACK_SGTL5000) \
     || defined(CONFIG_SND_SOC_IMX_3STACK_SGTL5000_MODULE)
+static int mxc_sgtl5000_plat_init(void);
+static int mxc_sgtl5000_plat_finit(void);
+static int mxc_sgtl5000_amp_enable(int enable);
 
 static struct mxc_sgtl5000_platform_data sgtl5000_data = {
 	.ssi_num = 1,
@@ -796,11 +799,13 @@ static struct mxc_sgtl5000_platform_data sgtl5000_data = {
 	.ext_port = 3,
 	.hp_irq = IOMUX_TO_IRQ(MX51_PIN_EIM_A26),
 	.hp_status = headphone_det_status,
-	.amp_gpo = "GPO2",
+	.amp_enable = mxc_sgtl5000_amp_enable,
 	.vddio = 1800000,
 	.vdda = 1800000,
 	.vddd = 12000000,
 	.sysclk = 12000000,
+	.init = mxc_sgtl5000_plat_init,
+	.finit = mxc_sgtl5000_plat_finit,
 };
 
 static struct platform_device sgtl5000_device = {
@@ -811,6 +816,41 @@ static struct platform_device sgtl5000_device = {
 	},
 };
 
+static int mxc_sgtl5000_plat_init(void)
+{
+	struct regulator *reg;
+	reg = regulator_get(&sgtl5000_device.dev, "GPO2");
+	if (IS_ERR(reg))
+		return -EINVAL;
+	sgtl5000_data.priv = reg;
+	return 0;
+}
+
+static int mxc_sgtl5000_plat_finit(void)
+{
+	struct regulator *reg;
+	reg = sgtl5000_data.priv;
+	if (reg) {
+		regulator_put(reg, &sgtl5000_device.dev);
+		sgtl5000_data.priv = NULL;
+	}
+	return 0;
+}
+
+static int mxc_sgtl5000_amp_enable(int enable)
+{
+	struct regulator *reg;
+	reg = sgtl5000_data.priv;
+
+	if (!reg)
+		return -EINVAL;
+	if (enable)
+		regulator_enable(reg);
+	else
+		regulator_disable(reg);
+	return 0;
+}
+
 static void mxc_sgtl5000_init(void)
 {
 	int err, pin;
-- 
1.5.4.4

